swarm_dial.go

```diff
@@ -343,71 +346,50 @@ func (s *Swarm) filterKnownUndialables(addrs []ma.Multiaddr) []ma.Multiaddr {
 	)
 }
 
-func (s *Swarm) dialAddrs(ctx context.Context, p peer.ID, remoteAddrs <-chan ma.Multiaddr) (transport.Conn, error) {
+func (s *Swarm) dialAddrs(ctx context.Context, p peer.ID, remoteAddrs <-chan ma.Multiaddr) (conn transport.Conn, err error) {
 	log.Debugf("%s swarm dialing %s", s.local, p)
 
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel() // cancel work when we exit func
 
-	// use a single response type instead of errs and conns, reduces complexity *a ton*
-	respch := make(chan dialResult)
+	dialResults := make(chan dialResult)
 
-	defaultDialFail := fmt.Errorf("failed to dial %s (default failure)", p)
-	exitErr := defaultDialFail
+	var lastDialErr error
 
 	defer s.limiter.clearAllPeerDials(p)
 
 	var active int
 	for remoteAddrs != nil || active > 0 {
-		// Check for context cancellations and/or responses first.
-		select {
-		case <-ctx.Done():
-			if exitErr == defaultDialFail {
-				exitErr = ctx.Err()
-			}
-			return nil, exitErr
-		case resp := <-respch:
-			active--
-			if resp.Err != nil {
-				log.Infof("got error on dial to %s: %s", resp.Addr, resp.Err)
-				// Errors are normal, lots of dials will fail
-				exitErr = resp.Err
-			} else if resp.Conn != nil {
-				return resp.Conn, nil
-			}
-
-			// We got a result, try again from the top.
-			continue
-		default:
-		}
-
```

Comment:
why was this code removed?

Reply:
Because it makes assumptions about timing that aren't worthwhile (instead there's just a lot of noisy code duplication). If it's a strict requirement that (which it isn't) that dial results be consumed before dials, the correct concurrency mechanisms should be used.

Reply:
I think we were doing this because dial might have finished before we started the next one. There is no guarantee of order you will receive messages in the select block.

You might first get all messages from `remoteAddrs` before consuming any from `respch` or context.

This is a use of concurrency mechanism.

Reply:
Which means, even if you've got a valid connection or context cancellation, without this additional loop we might continue dialling, which is just wasteful.

Reply:
@anacrolix please reinstate this code.

Reply:
In fact, this was added recently precisely to drain finished dials before starting new ones, thus making a better use of resources: https://github.com/libp2p/go-libp2p-swarm/pull/96

Reply:
Did it actually have a measurable effect? As I mentioned above, there are better alternative concurrency mechanisms that would have a more reliable impact. The "prioritized" select changed in #96 is optimistic at best. There's still a race between the first and second select, and any benefit relies on a whole bunch of assumptions about timings.

Reply:
There's always going to be a race here and there's nothing we can do about it, regardless of the synchronization mechanism. However, the extra select means that we'll start at most one unnecessary dial. Without it, we could end up draining `remoteAddrs` before we stop dialing.

Reply:
I ran the code with panics in place to detect when dials had actually completed and we pushed another dial job. It was exceptionally rare, and even then sits on the dial queue before being cancelled out. I've rehashed the entire function to make the priorities explicity without duplicating the dial response processing, I don't know if it's any better. Please take a look.